ca/ca.go

```diff
@@ -113,48 +119,55 @@ func (ca *CAImpl) makeRootCert(
 	return newCert, nil
 }
 
-func (ca *CAImpl) newRootIssuer() error {
+func newRootIssuer(ca *CAImpl) (*issuer, error) {
 	// Make a root private key
 	rk, err := makeKey()
 	if err != nil {
-		return err
+		return nil, err
 	}
 	// Make a self-signed root certificate
-	rc, err := ca.makeRootCert(rk, rootCAPrefix, nil)
+	rc, err := makeRootCert(ca, rk, rootCAPrefix, nil)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	ca.root = &issuer{
+	ca.log.Printf("Generated new root issuer with serial %s\n", rc.ID)
+	return &issuer{
 		key:  rk,
 		cert: rc,
-	}
-	ca.log.Printf("Generated new root issuer with serial %s\n", rc.ID)
-	return nil
+	}, nil
 }
 
-func (ca *CAImpl) newIntermediateIssuer() error {
-	if ca.root == nil {
-		return fmt.Errorf("newIntermediateIssuer() called before newRootIssuer()")
-	}
-
-	// Make an intermediate private key
-	ik, err := makeKey()
-	if err != nil {
-		return err
+func newIntermediateIssuer(root *issuer, ca *CAImpl, ik crypto.Signer) (*issuer, error) {
```

Comment:
I think that `newIntermediateIssuer` should generate its `ik crypto.Signer` itself rather than take a parameter. Since its always randomly generated in `newChain`. That also matches what `newRootIssuer` is doing and what `newIntermediateIssuer` used to do.

Sorry I missed that in my first pass.

Reply:
Actually, all intermediate certs must use the same private key, otherwise they aren't all valid intermediates for the leaf certificate. Maybe that's why you missed it in your first pass :)

Reply:
:man_facepalming: Good point! That makes sense :-)